packages/block-library/src/heading/heading-level-toolbar.js

```diff
@@ -0,0 +1,69 @@
+/**
+ * WordPress dependencies
+ */
+import { Toolbar, ToolbarGroup } from '@wordpress/components';
+import { __, sprintf } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import HeadingLevelIcon from './heading-level-icon';
+
+const HEADING_LEVELS = [ 1, 2, 3, 4, 5, 6 ];
+
+const POPOVER_PROPS = {
+	className: 'block-editor-heading-level-toolbar',
+	isAlternate: true,
+};
+
+/** @typedef {import('@wordpress/element').WPComponent} WPComponent */
+
+/**
+ * HeadingLevelToolbar props.
+ *
+ * @typedef WPHeadingLevelToolbarProps
+ *
+ * @property {number}   selectedLevel  The chosen heading level.
+ * @property {Function} onChange       Callback to run when toolbar value is changed.
+ */
+
+/**
+ * Toolbar for selecting a heading level (1 through 6).
+ *
+ * @param {WPHeadingLevelToolbarProps} props Component props.
+ *
+ * @return {WPComponent} The toolbar.
+ */
+export default function HeadingLevelToolbar( { selectedLevel, onChange } ) {
+	return (
+		<Toolbar
+			isCollapsed={ true }
+			icon={ <HeadingLevelIcon level={ selectedLevel } /> }
+			label={ __( 'Change heading level' ) }
+			popoverProps={ POPOVER_PROPS }
+		>
+			{ () => (
```

Comment:
@gziolo @diegohaz is this the correct way to use Toolbar and ToolbarGroup?

Reply:
The wrapper in this case should be a `ToolbarGroup` because it’s meant to be contained in a `role=“toolbar” element (the block toolbar).

From my understanding, the `ToolbarGroup` within the popover should be a sort of `DropdownMenu` instead. But it’s confusing to me that it’s horizontal, and not vertical as the other toolbar controls.

I see that it was designed vertically at first. I’m not sure how I feel with it looking similar to the control in the inspector, and not to the other toolbar controls.

Reply:
@diegohaz I used `Toolbar` because that's what `BlockAlignmentToolbar` uses. Notably, `Toolbar` basically just acts as a wrapper for `ToolbarGroup` anyway:

```js
function Toolbar( { className, __experimentalAccessibilityLabel, ...props } ) {
	if ( __experimentalAccessibilityLabel ) {
		return (
			<ToolbarContainer
				// `ToolbarGroup` already uses components-toolbar for compatibility reasons
				className={ classnames(
					'components-accessible-toolbar',
					className
				) }
				accessibilityLabel={ __experimentalAccessibilityLabel }
				{ ...props }
			/>
		);
	}

	return <ToolbarGroup { ...props } className={ className } />;
}
```

`ToolbarContainer` contains a nested `Toolbar` which end up rendering a `ToolbarGroup`.

So is `BlockAlignmentToolbar` doing it wrong, and should I just use `ToolbarGroup` directly?

Also, with regard to using `DropdownMenu`, it's worth noting that I think an inline warning for invalid heading levels should appear in the dropdown in the future. Does it still make sense to use `DropdownMenu` in that context? (It's also worth noting that my `isCollapsed` `Toolbar` is already using `Dropdown`.

The relation between all these different components is really confusing. :confounded:

Reply:
Once `__experimentalAccessibilityLabel` is stable, using `Toolbar` as a `ToolbarGroup` (that is, without passing an `accessibilityLabel` prop) will be deprecated.

And, conceptually speaking, having nested toolbars (`role="toolbar"`) doesn't make sense.

`BlockAlignmentToolbar` hasn't been migrated yet. It's not a big deal though as the migration is simple (just renaming `Toolbar` -> `ToolbarGroup`).

I totally agree that it's confusing, especially when you take into consideration the various possibilities of popovers (`Popover`, `Dropdown`, `DropdownMenu`, `isCollapsed`, `controls` or `children`).

Reply:
The whole `isCollapsed` thing is especially confusing. It takes a toolbar group and then creates a popover around it with a toggle to open the popover, while also changing the appearance of the toolbar to use visible text labels for every option in the toolbar. If a group of controls is hidden in a popover like that, are they still part of the parent toolbar (and therefore a "toolbar group"), or are they actually a separate toolbar (something that should have its own role) accessed by pressing a button in another toolbar?

I get the feeling that the `isCollapsed` prop was a questionable design choice since whether or not it's toggled drastically alters what kind of component is actually being rendered.

There's also the aforementioned idea of putting an inline heading level warning in the popover (similar to the color contrast warning for color controls). At that point, is the heading level thing still just a toolbar group, or is it a whole panel?

Anyway, I've pushed some changes to the implementation based on your feedback. Let me know what you think.